signature/provider.go

```
@@ -126,69 +117,58 @@ func (p *externalProvider) describeKey(ctx context.Context) (*plugin.DescribeKey
 	}
 
 	// Check keyID is honored.
 	if req.KeyID != resp.KeyID {
-		return nil, fmt.Errorf("keyID in generateSignature response %q does not match request %q", resp.KeyID, req.KeyID)
+		return nil, nil, fmt.Errorf("keyID in generateSignature response %q does not match request %q", resp.KeyID, req.KeyID)
 	}
 
-	// TODO: do we still need cert chain in response?
-	if _, err = parseCertChain(resp.CertificateChain); err != nil {
-		return nil, err
+	var certs []*x509.Certificate
+	if certs, err = parseCertChain(resp.CertificateChain); err != nil {
+		return nil, nil, err
 	}
-
-	return resp.Signature, nil
+	return resp.Signature, certs, nil
 }
 
-func (p *externalProvider) keyInfo() (signature.KeySpec, []*x509.Certificate, error) {
+// KeySpec return s keyspec of a keyID
+func (p *externalProvider) KeySpec() (signature.KeySpec, error) {
+	if p.keySpec != nil {
+		return *p.keySpec, nil
+	}
 	keyResp, err := p.describeKey(context.Background())
 	if err != nil {
-		return signature.KeySpec{}, nil, err
+		return signature.KeySpec{}, err
 	}
 
 	// Check keyID is honored.
 	if p.keyID != keyResp.KeyID {
-		return signature.KeySpec{}, nil, fmt.Errorf("keyID in describeKey response %q does not match request %q", keyResp.KeyID, p.keyID)
+		return signature.KeySpec{}, fmt.Errorf("keyID in describeKey response %q does not match request %q", keyResp.KeyID, p.keyID)
 	}
-	certs, err := parseCertChain(keyResp.CertificateChain)
-	if err != nil {
-		return signature.KeySpec{}, nil, err
-	}
-	return ParseKeySpecFromName(keyResp.KeySpec), certs, nil
-}
-
-// CertificateChain returns cert chain of a keyID
-func (p *externalProvider) CertificateChain() ([]*x509.Certificate, error) {
-	return p.certs, nil
-}
-
-// KeySpec return s keyspec of a keyID
-func (p *externalProvider) KeySpec() (signature.KeySpec, error) {
-	return p.keySpec, nil
+	keySpec := ParseKeySpecFromName(keyResp.KeySpec)
```

seems like `ParseKeySpecFromName` will return empty keySpec when `keyResp.KeySpec` is unsupported. Do you think we should return err instead of empty keySpec in this case?

---

How about still returning an empty keySpec. If `ParseKeySpecFromName ` returns an empty keySpec, `KeySpec` will call `describleKey`, and we can change keySpec to a struct type instead of a pointer.
```
if p.keySpec != signature.KeySpec{} {
      return p.keySpec, nil
}
keyResp, err := p.describleKey(context.Background())
// ...
p.keySpec = ParseKeySpecFromName(keyResp.KeySpec)
```